specs/AppLifecycle/Activation/AppLifecycle Activation.md

```diff
@@ -0,0 +1,506 @@
+the UWP optional attributes are very rarely used, and in some cases there are no apps in the Store
+that use a given parameter. For v1, the platform will support all the registry-based optional
+attributes - further UWP attributes may be included in a later release.
+
+The new `RegisterForXXXActivation` methods will be exposed as static methods from a new
+ActivationRegistrationManager class in the AppLifecycle component.
+
+The aim is that the app would call any of these RegisterForXXXActivation methods at any time
+dynamically from the app itself. For v1, this API does not support calling from the app's installer.
+For filetype and protocol, the app can call the RegisterForXXXActivation method multiple times, each
+time providing different filetype names or different protocol names. These registrations would be
+cumulative, so long as the parameters are different. If the app tries to register for the same
+filetype name or the same protocol multiple times, this is idempotent (there's only one registration
+per unique filetype name or protocol name; no error for multiple identical registrations, any new
+values supplied for the same registration overwrite any old values).
+
+Similarly, if the app registers multiple times for startup, this is also idempotent (a single
+registration, no error, any new values supplied overwrite any old values).
+
+```idl
+namespace Microsoft.Windows.AppLifecycle
+{
+    static runtimeclass ActivationRegistrationManager
+    {
+        static void RegisterForFileTypeActivation(String[] supportedFileTypes, String logo,
+            String displayName, String[] supportedVerbs, String exePath);
+
+        static void RegisterForProtocolActivation(String scheme, String logo, String displayName,
+            String exePath);
+
+        static void RegisterForStartupActivation(String taskId, String exePath);
+
+        static void UnregisterForFileTypeActivation(String[] fileTypes, String exePath);
+        static void UnregisterForProtocolActivation(String scheme, String exePath);
+        static void UnregisterForStartupActivation(String taskId);
+    }
+}
+```
+
+**RegisterForFileTypeActivation** registers the app for activation based on file types (extensions.)
+
+-   `supportedFileTypes` - one or more supported filetypes, specified by the file extension
+    including the leading ".", eg ".docx"
+-   `logo` - used by the Shell and Settings for this filetype. For packaged apps, a package-relative
+    path to an image file. For unpackaged, a literal filepath to a binary file (DLL, EXE) plus a
+    resource index. The caller can pass either.
+-   `displayName` - used by the Shell and Settings for this filetype.
+-   `supportedVerbs` - zero or more app-defined verbs. Each verb is added to the File Explorer
+    context menu when a registered file is right-clicked; and the selected verb is passed to the app
+    as the `FileActivatedEventArgs::Verb` property.
```

Comment:
What's the localization story? Can I pass a resource indirect string for automatic localization based on user preferred language?

For Win32 we generally distinguish between the display name of the verb (localized, shown to the user) and the programmatic name of the verb (not localized, passed to the app). This avoids apps having to do locale-sensitive parsing code when parsing their verbs. Particularly troublesome when words collide between languages.

```cpp
auto verb = args.Verb();
if (verb == L"spend" /* English */ || verb == L"bezahl" /* German */)
{
   DoSpending(); // German "spend" accidentally ends up doing this instead of DoDonate
}
else if (verb == L"donate" /* English */ || verb == L"spend" /* German */)
{
  DoDonate();
}
```

Reply:
Localization is an exercise for vNext, we cannot fit it into v1.